Add Warning methods onto Any to make working with them easier.

jdunkerley opened the pull request:
### Pull Request Description

Adds `has_warnings`, `remove_warnings` and `throw_on_warning` extension methods.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

jdunkerley committed:
Add tests and trivial implementations (doesn't work).

hubertp committed:
Introduce Warning-related methods to Any

This change adds a special method dispatch to values with warnings.
That way one can call methods like `has_warnings`, `remove_warnings` or
`throw_on_warning` on any value and get the expected results.

The change requires a special handling in method dispatch because of the
way warnings are being propagated during execution - we unwrap and then
append warnings during method calls.
One additional problem here was that methods on Warning builtin type are
part of the eigentype and have a synthetic `self` parameter, which would
make calls **not** fully applied unless we add that (useless but still)
argument during method dispatch.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
+            .lookupMethodDefinition(
+                EnsoContext.get(this).getBuiltins().warning().getEigentype(), symbol.getName());
+    if (function != null) {
+      // Warning's builtin type methods are static meaning that they have a synthetic `self`
```
@JaroslavTulach Would very much appreciate your feedback on this bit. Unfortunately due to synthetic `self` inserted for static methods I have to somehow convince runtime that it is OK to call this function and it is fully applied. I figured that we are dealing with Warnings it's OK to hit this slow path.

JaroslavTulach replied:
Where is the slow path? Only when creating the node? That's obviously have to be slow path. However it'd be better to have this AST manipulation be done automatically by the Truffle DSL.

jdunkerley committed:
Add more tests and update `remove_warnings`.

jdunkerley committed:
CHANGELOG and missing tests.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
+      Object[] arguments,
+      @Cached MethodResolverNode methodResolverNode,
+      @CachedLibrary(limit = "10") WarningsLibrary warnings) {
+    Function function =
+        symbol
+            .getScope()
+            .lookupMethodDefinition(
+                EnsoContext.get(this).getBuiltins().warning().getEigentype(), symbol.getName());
```
Do I understand correctly that we modify the dispatch mechanism, so that if a value `x` has warnings attached, then for a call `x.f` we will first try to resolve any methods `f` defined on the `Warning` type and only later will fall back to the actual type of `x`?

This is a highly non trivial modification of the dispatch mechanism. We currently use it in an innocent way, because the relevant methods are defined both on `Warning` and on `Any` and we don't expect any overrides.

But I'm suspicious - since there exists a method `Warning.get_all`, if I do:
```
type My_Fancy_Collection
    Value (x:Integer)

    get_all : Vector Integer
    get_all self = [self.x]

main =
    a = My_Fancy_Collection.Value 42
    b = Warning.attach "WARN" <| My_Fancy_Collection.Value 23

	IO.println a.get_all
	IO.println b.get_all
```
will I get
```
[42]
[23]
```
as I'd expect or will I get some unclear behaviour due to `Warning.get_all` shadowing my `get_all`?

Please add a test for this case.

radeusgd replied:
I really don't like completely overriding the dispatch mechanism if a value has warnings attached. I don't expect a value with warnings to behave much differently than a regular one.

I don't want to impose any solutions, but maybe instead of redirecting 'all' calls dispatched on a value with warnings through the `Warning` type (which as I noted above may have potentially bad implications), we can just have a list of these special methods? I'd even suggest just creating a map of `methods that need to have warnings from self passed to it`. That would make it a bit less extensible but at the same time more clear which methods are special - and IMO in this case the clarity is the more important goal here.

radeusgd replied:
I think we should _definitely_ add a test for the `get_all` case, regardless of the chosen solution. Especially as `get_all` is a pretty common name and the possibility of changing a method declared by the user to one on the warnings module if warnings are attached is something to be avoided.

Then, if we choose to stay with the changed dispatch mechanism, I think it's very important to have some piece of documentation for it. Even if just a comment in the docstring of the `InvokeMethodNode` - but some description that can be found if we ever wonder why dispatch on values-with-warnings behaves the way it does.

radeusgd reviewed (changes requested):
The API and behaviour (tests) look all good.

But I'm quite worried that the modification to the dispatch mechanism when warnings are attached may have some unclear implications. I think we need to document this and add the mentioned `get_all` test (which is my main worry here).

radeusgd commented:
After discussing with @jdunkerley we agreed that dispatching all methods to the `Warning` type if a value has warnings attached may potentially lead to unexpected behaviour.

It may be better to instead just have a list of 'special' methods that need to have the special handling of warnings on `self` being preserved. That should make it more efficient and it will be also more explicit which methods have this special handling. We should probably document on the `Warning` type where in the runtime one can add such methods.

Then we realised we may actually need this behaviour to be more generalized and work not only with the `Warning` type. Actually, for the **Write** capabilities, we want by default to avoid writing if the value contains any warnings. To be able to effectively stop a `table.write some args` operation if `table` has warnings attached, we will need `Table.write` to also have an ability to see any warnings attached to `self` (and due to how CB is designed, we want to keep this dispatched as `table.write some args` and not `Some_Writer_Type.write table some args`).

So ideally, we should have some way to mark _any_ method to be able to have this special handling of attached warnings being propagated into the `self` instance. Possibly, we could use some annotation to do that, but other solutions may be explored.

The plan is to:
1. Now do the simple fix-up and use the list of special methods for now.
2. Plan to create the proper solution allowing us to mark any method as needing this special warning handling, as soon as reasonably possible - as it will be needed for Write anyway.
   #6222

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
+            .getScope()
+            .lookupMethodDefinition(
+                EnsoContext.get(this).getBuiltins().warning().getEigentype(), symbol.getName());
+    if (function != null) {
```
I think the `function` "searching" shall be put into its own method, it should be accepted as `@Cached FunctionSchema` argument and used as a guard for this `@Specialization`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
+      } else {
+        arguments1 = arguments;
+      }
+      if (warningFunctionNode == null) {
```
`warningFunctionNode` shall also be accepted as `@Cached` argument - e.g. when the specialization is triggered, you'll get the node and you don't have to mangle with the AST manually.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
+      // and call including an additional, dummy, argument.
+      Object[] arguments1;
+      if (!function.getSchema().isFullyApplied()) {
+        arguments1 = new Object[function.getSchema().getArgumentsCount()];
```
`function.getSchema()` shall be cached, to be compilation constant.

hubertp commented:
@radeusgd I'm a bit allergic to hardcoding method names in the compiler as that has a tendency to quickly get out of sync with what libraries define and adds more magic to what is happening.

Instead, what I would propose is that when doing a method dispatch on a value with warnings, we check where the function, that will be invoked, comes from come. If it is `Any` or `Warning` we proceed as defined currently in the PR. If it is some custom type, we default to the old dispatch. That would make your example pass.

Regarding the second requirement, I will leave the discussion in the corresponding ticket.

JaroslavTulach reviewed (approved):
If possible, replace the manual AST manipulation with a Truffle DSL. I believe it is possible to have a `@Cached` `Node`. If so, then that's the preferred variant.

radeusgd commented:
> @radeusgd I'm a bit allergic to hardcoding method names in the compiler as that has a tendency to quickly get out of sync with what libraries define and adds more magic to what is happening.
> 
> Instead, what I would propose is that when doing a method dispatch on a value with warnings, we check where the function, that will be invoked, comes from come. If it is `Any` or `Warning` we proceed as defined currently in the PR. If it is some custom type, we default to the old dispatch. That would make your example pass.

Okay, that sounds ok although it does complicate the dispatch mechanism quite a bit.

I'm not a fan of hardcoding names either, I just thought of it as a way to contain the 'weirdness' of this method dispatch change into as small set of functions as possible.

But I think the idea to only do special dispatch if the method is defined on both `Warning` and `Any` sounds ok. Let's just make sure this irregularity is well documented.

Also, can we do so this applies only to non-static methods? I.e. `Warning.get_all` would not make a `value_with_warnings.get_all` work, unless we define `get_all` as **non**-static.

Why am I asking for this? Because I'd like to minimize the amount of irregularity. IMO this should only apply to warning-related methods that are defined **both** on `Any` and `Warning`, so that they will work on all values predictably.

I definitely would not want to have `x.get_all` fail with unresolved method when there are no warnings and return a vector if there are some. It should either always return the vector or always fail with unresolved.

Maybe we could add a test that gets all these 'specially dispatched' methods on `Warning` and verifies that there are corresponding counterparts on `Any`?

I know this is getting complicated, that's why I was advocating for landing the hardcoded names as a quick-fix and then doing #6222 as follow-up. But I think it's really important to avoid making method dispatch irregular and hard to understand - it's one of the most basic primitives of the language.

hubertp committed:
Don't override dispatch on custom types

Added a guard that makes sure we still can define methods from Warning
on custom types and they will be correctly dispatched.

hubertp commented:
@radeusgd does my last change satisfy your requirement (minus the missing documentation for the special behaviour)?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -145,50 +147,25 @@ Object doWarning(
       Object self,
       Object[] arguments,
-      @Cached MethodResolverNode methodResolverNode,
+      @CachedLibrary(limit = "10") TypesLibrary dispatch,
```
Why is the `TypesLibrary` called `dispatch` here? I was wondering what is `dispatch.getType` becuase the name `dispatch` suggested to me this is some special `getType` used for dispatch, not the standard type library one.

hubertp replied:
Following the `doFunctionDispatch` convention. I can change it to `types`

radeusgd replied:
I see. IMHO `types` is much more readable, but I don't want to be stubborn.

radeusgd reviewed (approved):
Yeah, I think it looks ok.

hubertp committed:
Add specialization for custom warning dispatch

Rather than employing the desired logic via node manipulation, we use
Truffle's DSL to achieve the equivalent semantics, as per PR review.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +145,105 @@ Object doPanicSentinel(
+      throw new IllegalStateException(e);
+    }
+
+    Type typeOfSymbol = symbol.resolveTypeFor(types.getType(selfWithoutWarnings));
```
`resolveDeclaringType`, maybe? Just like [getDeclaringClass](https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Member.html#getDeclaringClass())?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +145,105 @@ Object doPanicSentinel(
+      @CachedLibrary(limit = "10") WarningsLibrary warnings,
+      @Cached("resolveWarningFunction(self, symbol, types, warnings)") Function resolvedFunction,
+      @Cached("resolvedFunction.getSchema()") FunctionSchema cachedSchema,
+      @Cached(value = "argumentsWithExplicitSelf(cachedSchema, arguments)", dimensions = 1)
```
How can you cache arguments? `arguments` are going to be different with every invocation. Caching something based on their value (even without a guard!) doesn't seem to be good idea.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -134,13 +145,105 @@ Object doPanicSentinel(
+  }
+
+  CallArgumentInfo[] callArgumentsWithExplicitSelf() {
+    int length = invokeFunctionNode.getSchema().length;
```
Caching call info based on `invokeFunctionNode` field? That is very suspicious. How that is supposed to work?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -164,15 +267,23 @@ Object doWarning(
+      arrOfWarnings = warnings.getWarnings(self, this);
+    } catch (UnsupportedMessageException e) {
+      // Not possible due to `hasWarnings` check
+      throw new IllegalStateException(e);
```
`shouldNotReachHere` allows one to add the comment as a string to parameter to the method.

hubertp replied:
We are not particularly consistent on that and throw `IllegalStateException` on such occasions. But sure, I can change it

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -405,6 +529,7 @@ Object doConvertZonedDateTime(
       Object[] arguments,
       @CachedLibrary(limit = "10") TypesLibrary types,
       @CachedLibrary(limit = "10") InteropLibrary interop,
+      @CachedLibrary(limit = "10") WarningsLibrary warnings,
```
What is this good for? I don't see the `warnings` used anywhere...

hubertp replied:
It is used, in the guard

JaroslavTulach reviewed (changes requested):
I am not sure I fully understand the caching & specialization. Few questions inline.

hubertp committed:
nits

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -164,15 +271,23 @@ Object doWarning(
     }
 
-    arguments[thisArgumentPosition] = self.getValue();
-    ArrayRope<Warning> warnings = self.getReassignedWarnings(this);
```
So we are getting rid of the reassigned warnings altogether? Are we still going to be able to find out the origin of the warning?

hubertp replied:
`getWarnings` does the reassignment AFAIK https://github.com/enso-org/enso/blob/develop/engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java#L118

hubertp committed:
s/IllegalStateException/shouldNotReachHere

hubertp committed:
Cosmetic changes and fix native-image

Some simplifications after discussing with Jaroslav.

hubertp committed:
Bring back IllegalStateException

Looks like there is a bug in Graal that prevents us from using
`shouldNotReachHere`.
